connector/src/main/scala/com/microsoft/kusto/spark/datasource/KustoFilter.scala

```diff
@@ -0,0 +1,122 @@
+package com.microsoft.kusto.spark.datasource
+
+import java.sql.{Date, Timestamp}
+
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+
+private[kusto] object KustoFilter {
+  // Augment the original query to include column pruning and filtering
+  def pruneAndFilter(originalSchema: StructType, originalQuery: String, filtering: KustoFiltering): String = {
+    var query = originalQuery
+
+    if (!filtering.filters.isEmpty) {
+      query += KustoFilter.buildFiltersClause(originalSchema, filtering.filters)
+    }
+
+    if (!filtering.columns.isEmpty) {
+      query += KustoFilter.buildColumnsClause(filtering.columns)
+    }
+
+    query
+  }
+
+  def pruneSchema(schema: StructType, columns: Array[String]): StructType = {
+    val fieldMap = Map(schema.fields.map(x => x.name -> x): _*)
+    new StructType(columns.map(name => fieldMap(name)))
+  }
+
+  def buildColumnsClause(columns: Array[String]): String = {
+    if(columns.isEmpty) "" else {
+      " | project " + columns.mkString(", ")
+    }
+  }
+
+  def buildFiltersClause(schema: StructType, filters: Seq[Filter]): String = {
+    val filterExpressions = filters.flatMap(f => buildFilterExpression(schema, f)).mkString(" and ")
+    if (filterExpressions.isEmpty) "" else " | where " + filterExpressions
+  }
+
+  def buildFilterExpression(schema: StructType, filter: Filter): Option[String] = {
+
+    filter match {
+      case EqualTo(attr, value) => binaryScalarOperatorFilter(schema, attr, value, "==")
+      case EqualNullSafe(attr, value) if value == null => unaryScalarOperatorFilter(attr, "isnull")
+      case EqualNullSafe(attr, value) => binaryScalarOperatorFilter(schema, attr, value, "==")
+      case GreaterThan(attr, value) => binaryScalarOperatorFilter(schema, attr, value, ">")
+      case GreaterThanOrEqual(attr, value) => binaryScalarOperatorFilter(schema, attr, value, ">=")
+      case LessThan(attr, value) => binaryScalarOperatorFilter(schema, attr, value, "<")
+      case LessThanOrEqual(attr, value) => binaryScalarOperatorFilter(schema, attr, value, "<=")
+      case In(attr, values) => unaryOperatorOnValueSetFilter(schema, attr, values, "in")
+      case IsNull(attr) => unaryScalarOperatorFilter(attr, "isnull")
+      case IsNotNull(attr) => unaryScalarOperatorFilter(attr, "isnotnull")
+      case And(left, right) => binaryLogicalOperatorFilter(schema, left, right, "and")
+      case Or(left, right) => binaryLogicalOperatorFilter(schema, left, right, "or")
+      case Not(child) => unaryLogicalOperatorFilter(schema, child, "not")
+      case StringStartsWith(attr, value) => stringOperatorFilter(schema, attr, value, "startswith_cs")
+      case StringEndsWith(attr, value) => stringOperatorFilter(schema, attr, value, "endswith_cs")
+      case StringContains(attr, value) => stringOperatorFilter(schema, attr, value, "contains_cs")
```

Comment:
we should consider using 'has' which is much much faster but is only for whole terms capturing

Reply:
we can also consider some configuration settings for the case sensitivity

Reply:
"has" - since it is really a string comparison (and not necessarily a whole entity) I don't see how I can use it. 
Case sensitivity - I'll add a task, maybe we'll add as a configuration option